### PR TITLE
0x/cli set soft limit tx size

### LINKED ISSUE #6411: Blob tx batches bigger than 1 blob
### Describe the feature

Recent improvements to `TransactionManager` included batching tx imports to commit to the pool. For blob txns requested over an eth68 connection, this will be trivial. This is due to the fact that the soft limit for the expected response size, with respect to which `GetPooledTransaction` are filled with hashes, is 128 KB. That means each request will only request one blob tx. Either the expected response size soft limit is increased, devp2p spec's standard size limit for responses 2 MiB would be great, or batches need to persist the scope of `import_transactions`, in order to allow filling the batch up with a non-trivial  number of elements before committing to the pool, when the limit of 128 KB is kept.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by justcode740**
Closes https://github.com/paradigmxyz/reth/issues/6411.

Based on https://github.com/paradigmxyz/reth/issues/6411#issuecomment-1932436372, i try to make DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE a cli-set value.

Based on both https://github.com/paradigmxyz/reth/issues/6411#issuecomment-1932436372 and https://github.com/paradigmxyz/reth/pull/6519#discussion_r1484929778, i keep the name `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` instead of `SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` as in the based branch.

currently the value hasn't been set from the cli param, i have some questions:
1 should `soft_limit_byte_size_pooled_transactions_response` be part of network config
2. best way to port over the value to `transactionsmanager`? atm when `transactionsmanager` is constructed, seem neither network config nor cli params are in scope
https://github.com/paradigmxyz/reth/blob/71fddff6b1ae184d6e89ac69ae67dafb1f78dd97/crates/net/network/src/builder.rs#L62

the other place transactionmanager is constructed but guess this one can just use the default const value for testing purpose:
https://github.com/paradigmxyz/reth/blob/71fddff6b1ae184d6e89ac69ae67dafb1f78dd97/crates/net/network/src/test_utils/testnet.rs#L393

### CODE REVIEW

**Comment by justcode740** [comment]
ok, i guess i rebase bit too late...there're so many relevant changes in this branch https://github.com/paradigmxyz/reth/commits/emhane/speed-up-request-buffered-hashes/, wonder what's the diff between `SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` and `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_ASSEMBLE_GET_POOLED_TRANSACTIONS_REQUEST`? and is using such long const name normal?..

**Comment by justcode740** [comment]
> ok, i guess i rebase bit too late...there're so many relevant changes in this branch https://github.com/paradigmxyz/reth/commits/emhane/speed-up-request-buffered-hashes/, wonder what's the diff between `SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` and `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_ASSEMBLE_GET_POOLED_TRANSACTIONS_REQUEST`? and is using such long const name normal, tbh it's quite confusing to read and understand here.. @emhane

so should i change `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_ASSEMBLE_GET_POOLED_TRANSACTIONS_REQUEST` instead of `SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` now, i already assume `SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` is the one i should change and rename it to `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` but if it's `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_ASSEMBLE_GET_POOLED_TRANSACTIONS_REQUEST` then i can make the changes, the test currently fails also due to this, would pass if i change the size to 128 * 1024 instead of 2  * 1024 * 1024.

File: crates/net/network/src/transactions/fetcher.rs [hunk 1]
```diff
@@ -190,7 +194,7 @@ impl TransactionFetcher {
             hashes_to_request.push(hash);
 
             // tx is really big, pack request with single tx
-            if size >= DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_ASSEMBLE_GET_POOLED_TRANSACTIONS_REQUEST {
+            if size >= self.soft_limit_byte_size_pooled_transactions_response {
```
Comment (emhane) [inline crates/net/network/src/transactions/fetcher.rs:197]: here the correct constant has been replaced for example

File: crates/net/network/src/transactions/fetcher.rs [hunk 2]
```diff
@@ -51,6 +52,9 @@ pub(super) struct TransactionFetcher {
     pub(super) filter_valid_hashes: AnnouncementFilter,
     /// Info on capacity of the transaction fetcher.
     info: TransactionFetcherInfo,
+    /// Soft limit for the byte size of a [`PooledTransactions`](reth_eth_wire::PooledTransactions)
+    /// response.
+    pub(super) soft_limit_byte_size_pooled_transactions_response: usize,
```
Comment (emhane) [inline crates/net/network/src/transactions/fetcher.rs:57]:
```suggestion
    /// Soft limit for the byte size of the expected [`PooledTransactions`](reth_eth_wire::PooledTransactions)
    /// response on packing a [`GetPooledTransactions`] request with hashes.
    pub(super) soft_limit_byte_size_pooled_transactions_response_on_pack_request: usize,
```

Comment (emhane) [inline crates/net/network/src/transactions/fetcher.rs:57]: please move `soft_limit_byte_size_pooled_transactions_response` into `TransactionFetcherInfo`

File: crates/net/network/src/transactions/mod.rs [hunk 3]
```diff
@@ -310,7 +308,7 @@ where
             let transactions = self.pool.get_pooled_transaction_elements(
                 request.0,
                 GetPooledTransactionLimit::ResponseSizeSoftLimit(
-                    SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE,
+                    self.transaction_fetcher.soft_limit_byte_size_pooled_transactions_response,
```
Comment (emhane) [inline crates/net/network/src/transactions/mod.rs:311]:
```suggestion
                    SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE,
```

File: crates/net/network/src/transactions/mod.rs [hunk 4]
```diff
@@ -77,10 +77,8 @@ use constants::SOFT_LIMIT_COUNT_HASHES_IN_NEW_POOLED_TRANSACTIONS_BROADCAST_MESS
 use fetcher::{FetchEvent, TransactionFetcher};
 pub use validation::*;
 
-use self::constants::{
-    tx_manager::*, DEFAULT_SOFT_LIMIT_BYTE_SIZE_TRANSACTIONS_BROADCAST_MESSAGE,
-    SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE,
-};
+pub use self::constants::DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE;
+use self::constants::{tx_manager::*, DEFAULT_SOFT_LIMIT_BYTE_SIZE_TRANSACTIONS_BROADCAST_MESSAGE};
```
Comment (emhane) [inline crates/net/network/src/transactions/mod.rs:81]:
```suggestion
use self::constants::{
    tx_manager::*, DEFAULT_SOFT_LIMIT_BYTE_SIZE_TRANSACTIONS_BROADCAST_MESSAGE,
    SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE,
};
```

File: crates/transaction-pool/src/validate/constants.rs [hunk 5]
```diff
@@ -4,11 +4,11 @@
 /// O(maxslots), where max slots are 4 currently).
 pub const TX_SLOT_BYTE_SIZE: usize = 32 * 1024;
 
-/// [`MAX_TX_INPUT_BYTES`] is the maximum size a single transaction can have. This field has
+/// [`DEFAULT_MAX_TX_INPUT_BYTES`] is the default maximum size a single transaction can have. This field has
 /// non-trivial consequences: larger transactions are significantly harder and
 /// more expensive to propagate; larger transactions also take more resources
 /// to validate whether they fit into the pool or not.
```
Comment (emhane) [inline crates/transaction-pool/src/validate/constants.rs:10]:
```suggestion
/// to validate whether they fit into the pool or not. Default is 4 times [`TX_SLOT_BYTE_SIZE`], which defaults to 32 KiB, so 128 KiB.
```

File: crates/node-core/src/args/network_args.rs [hunk 6]
```diff
@@ -73,6 +76,11 @@ pub struct NetworkArgs {
     /// Maximum number of inbound requests. default: 30
     #[arg(long)]
     pub max_inbound_peers: Option<usize>,
+
+    /// Soft limit for the byte size of a [`PooledTransactions`](reth_eth_wire::PooledTransactions)
+    /// response.
+    #[arg(long = "pooled-tx-response-soft-limit", value_name = "BYTES", default_value_t = DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE)]
+    pub soft_limit_byte_size_pooled_transactions_response: usize,
 }
```
Comment (emhane) [inline crates/node-core/src/args/network_args.rs:84]: please add to the `help` for this arg, that this is a spec'd value that should only be set for experimental purposes on a testnet.

**Review by emhane — CHANGES_REQUESTED** [review]
Almost the right parameter, `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_ASSEMBLE_GET_POOLED_TRANSACTIONS_REQUEST` is the one we want to expose.

That said, we can leave the `SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` exposed in cli too, now that you have done the work. Mind here, as mentioned in another comment, the help for this cli arg needs to include that this is a spec'd value that should only be set for experimental purposes.

Please put both the `soft_limit_byte_size_pooled_transactions_response_on_pack_request` and `soft_limit_byte_size_pooled_transactions_response` in `TransactionFetcherInfo`. The field `soft_limit_byte_size_pooled_transactions_response` needs to have visibility `pub(super)` so `TransactionsManager` can read it.

File: crates/net/network/src/transactions/constants.rs [hunk 7]
```diff
@@ -25,7 +25,7 @@ pub const SOFT_LIMIT_COUNT_HASHES_IN_GET_POOLED_TRANSACTIONS_REQUEST: usize = 25
 /// request. Spec'd at 2 MiB.
 ///
 /// <https://github.com/ethereum/devp2p/blob/master/caps/eth.md#protocol-messages>.
-pub const SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE: usize = 2 * 1024 * 1024;
+pub const DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE: usize = 2 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (justcode740) [reply crates/net/network/src/transactions/constants.rs:28]: yeah, but it was 128 before changes made in the target branch, guess i rebase bit too late.

File: crates/net/network/src/transactions/fetcher.rs (hunk 2 above)
Comment (justcode740) [reply crates/net/network/src/transactions/fetcher.rs:57]: u mean `soft_limit_byte_size_pooled_transactions_response_on_pack_request` right, also `soft_limit_byte_size_pooled_transactions_response_on_pack_request`'s default value is `SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE` which spec'd at 2 MiB? could u point me where it's spec'd?

File: crates/net/network/src/transactions/fetcher.rs (hunk 1 above)
Comment (justcode740) [reply crates/net/network/src/transactions/fetcher.rs:197]: this is actually incorrect, should be `soft_limit_byte_size_pooled_transactions_response_on_pack_request`

File: crates/net/network/src/transactions/constants.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (justcode740) [reply crates/net/network/src/transactions/constants.rs:28]: also why this `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_ASSEMBLE_GET_POOLED_TRANSACTIONS_REQUEST` not name as `DEFAULT_SOFT_LIMIT_BYTE_SIZE_POOLED_TRANSACTIONS_RESPONSE_ON_PACK_GET_POOLED_TRANSACTIONS_REQUEST`. anyway will follow your naming here.

File: crates/net/network/src/transactions/mod.rs (hunk 4 above)
Comment (justcode740) [reply crates/net/network/src/transactions/mod.rs]: i'd need to use it to set the default value for cli args, so need to make it a public use?

**Comment by justcode740** [comment]
resolve comments above. put my leftover question here:

currently the value hasn't been set from the cli param, i have some questions:
1 should `soft_limit_byte_size_pooled_transactions_response` be part of network config
2. best way to port over the value to `transactionsmanager`? atm when `transactionsmanager` is constructed, seem neither network config nor cli params are in scope
https://github.com/paradigmxyz/reth/blob/71fddff6b1ae184d6e89ac69ae67dafb1f78dd97/crates/net/network/src/builder.rs#L62

the other place transactionmanager is constructed but guess this one can just use the default const value for testing purpose:
https://github.com/paradigmxyz/reth/blob/71fddff6b1ae184d6e89ac69ae67dafb1f78dd97/crates/net/network/src/test_utils/testnet.rs#L393

**Comment by justcode740** [comment]
> resolved comments above. put my leftover question here:
> 
> currently the value hasn't been set from the cli param, i have some questions: 1 should `soft_limit_byte_size_pooled_transactions_response` be part of network config 2. best way to port over the value to `transactionsmanager`? atm when `transactionsmanager` is constructed, seem neither network config nor cli params are in scope
> 
> https://github.com/paradigmxyz/reth/blob/71fddff6b1ae184d6e89ac69ae67dafb1f78dd97/crates/net/network/src/builder.rs#L62
> 
> the other place transactionmanager is constructed but guess this one can just use the default const value for testing purpose:
> 
> https://github.com/paradigmxyz/reth/blob/71fddff6b1ae184d6e89ac69ae67dafb1f78dd97/crates/net/network/src/test_utils/testnet.rs#L393

@emhane would appreciate some inputs on this, and get it merged soon.

File: crates/net/network/src/transactions/mod.rs (hunk 4 above)
Comment (emhane) [reply crates/net/network/src/transactions/mod.rs]: oh yeah, my bad

File: crates/net/network/src/transactions/fetcher.rs (hunk 1 above)
Comment (emhane) [reply crates/net/network/src/transactions/fetcher.rs]: that comment was left before that variable existed in the code. spot on it should be `soft_limit_byte_size_pooled_transactions_response_on_pack_request`.

**Comment by justcode740** [comment]
> can delet

oh, hey i just pushed the additional changes, wonder why is it closed? @emhane , was planning to do it yesterday but dragged into some personal event so bit delayed

**Comment by Rjected** [comment]
> > can delet
> 
> oh, hey i just pushed the additional changes, wonder why is it closed? @emhane , was planning to do it yesterday but dragged into some personal event so bit delayed

This needs to be rebased on main, since the `speed-up-request-buffered-hashes` branch has been merged

**Comment by justcode740** [comment]
> speed-up-request-buffered-hashes

yeah, i opened another pr here. https://github.com/paradigmxyz/reth/pull/6594

**Pull request closed without merging**